File: src/mat/TestSGP4.m
% verify sgp4 implementation by looping over verification output
% and simulated tles for corresponding mins since epoch

% count tles
fid = fopen('../../data/SGP4-VER.TLE');
tlecnt = 0;
line1 = fgetl(fid);
while ischar(line1)
    if(line1(1)=='1')
        tlecnt = tlecnt +1;
    end
    line1 = fgetl(fid);
end

fclose(fid);

% create tle cell array
tles = cell(tlecnt);

fid = fopen('../../data/SGP4-VER.TLE');

ind = 1;
line1 = fgetl(fid);
while ischar(line1)
    if(line1(1)=='1')
        line2 = fgetl(fid);
        tle = TLE(line1,line2);
        tles{ind} = tle;
        ind = ind+1;
    end
    line1 = fgetl(fid);
end

fclose(fid);

curobj = 0;
strind = 0;
ind = 0;
prevrv = 0;

% loop over verification file to compute rv for various tles and times
fid = fopen('../../data/tcppver.out');
line1 = fgetl(fid);
cnt = 0;
rerr = 0;
verr = 0;
while ischar(line1)
    strind = strfind(line1,'xx'); 
    if(strind > 0)
        %% get object number
        curobj = str2double(strtrim(line1(1:strind-1)));

        %% find tle with matching object number
        ind = 1;
        while(ind <= tlecnt && tles{ind}.objectNum ~= curobj)
            ind = ind+1;
        end
        tle = tles{ind};

    elseif(curobj > 0)

        %% get RV for specified number of minutes
        verout = sscanf(line1,'%f');
        rv = tle.getRV(verout(1));
        %% original code reused rv so if there is an error we need to carry old values
        if(tle.sgp4Error > 0)
            rv = prevrv;
        end

        rdist = sqrt( (rv(1)-verout(2))^2 + (rv(2)-verout(3))^2 + (rv(3)-verout(4))^2);
        vdist = sqrt( (rv(4)-verout(5))^2 + (rv(5)-verout(6))^2 + (rv(6)-verout(7))^2);
        cnt = cnt+1;
        rerr = rerr+rdist;
        verr = verr+vdist;
        if(rdist > 1e-7 || vdist > 1e-9)
            sprintf('%d\t%d\t%.15f\t%.15f\n',curobj,verout(1),rdist,vdist)
        end

        prevrv = rv;
    end

    line1 = fgetl(fid);
end
fclose(fid);
verr =verr/cnt;
rerr = rerr/cnt;

sprintf('Typical errors r=%E mm\tv=%E mm/s\n',(1e6*rerr),(1e6*verr))

File: src/python/TestSGP4.py
import json
from TLE import TLE
from math import sqrt

cnt = 0
rerr = 0
verr = 0

# compare the propagated TLE to entries
def check(tle, entries, lastrv):
    for entry in entries:
        lastrv = checkRV(tle,entry,lastrv)

    return lastrv

def checkRV(tle,entry,lastrv):
    min = entry['minutesSinceEpoch']
    rv = tle.getRV(min)

    if(tle.sgp4Error == 0):
        rdist = dist(rv[0],entry['r'])
        vdist = dist(rv[1],entry['v'])
    else:
        rdist = dist(lastrv[0],entry['r'])
        vdist = dist(lastrv[1],entry['v'])

    if(rdist > 1.2e-7 or vdist > 1e-8):
#        print tle.objectNum, min, rdist, vdist, tle.sgp4Error
# an attempt at making python2/3 independent printing
        print("\nError")
        print(tle.objectNum)
        print(min)
        print(rdist)
        print(vdist)
        print(tle.sgp4Error)

    global cnt
    global rerr
    global verr
    cnt = cnt+1
    rerr = rerr + rdist
    verr = verr + vdist
    return rv

# two-norm vec3
def dist(v1, v2):
    tmp = 0
    tot = 0 

    tmp = v1[0]-v2[0]
    tot += tmp*tmp
    tmp = v1[1]-v2[1]
    tot += tmp*tmp
    tmp = v1[2]-v2[2]
    tot += tmp*tmp

    return sqrt(tot)



# Read JSON
verinlist = 0 
veroutlist = 0

# read tles
with open('../../data/sgp4-ver.json') as json_file:  
    verinlist = json.load(json_file)

# read output
with open('../../data/sgp4-ver.out.json') as json_file:  
    veroutlist = json.load(json_file)

# loop over tles
ind = 0
verout = 0
lastrv = [[0,0,0],[0,0,0]]
for verin in verinlist:
    verout = veroutlist[ind]
    tle = TLE(verin['line1'],verin['line2'])

    # the original verification had a loop where rv were not reset, 
    # so in case of an error we need the previous rv
    lastrv = check(tle,verout['entries'],lastrv)

    ind=ind+1

rerr = rerr/cnt
verr = verr/cnt
print("Typical Errors")
print(1e6*rerr)
print("mm")
print(1e6*verr)
print("mm/s")
